### ISSUE TITLE
RPN fg/bg imbalance

### ISSUE THREAD

**Issue by korkinof** [rbgirshick/py-faster-rcnn#144]
Hi guys,
This approach is highly appealing for us and  I have managed to build a similar codebase in Torch.
However, I am having some trouble training in Torch and lately I have been comparing with the Caffe implementation which seems to train normally. I have debugged and printed several metrics to figure out what is happening.
As a result of my investigation I have observed the following:
In our dataset the Stage 1 RPN training yields too few foreground bounding boxes. Around 7% FG and 93% BG. That is of course because the default scales&ratios are not very suitable for the objects we have, but...
The Torch implementation gets all samples to the background class and gets stuck there. In the case of Caffe and after some time the foreground accuracy starts rising as well and quite rapidly.
My question is if there is any per-class normalisation taking place in the Caffe loss layer or re-weighting or something like that.
I cannot exclude the possibility of a having a bug in my implementation, however I have had the same issue with imbalanced datasets before and in other models as well (i.e. Random Forests). The solution was always normalising to account for the imbalance, but I could not find anything like that in the Caffe layer implementation.
The other possible difference I could think of is in the learning rate, which may be too high or too low in Torch.
Can anyone shed some light on that? I would really appreciate it!
Best,
Dimitrios

**Comment by korkinof** [comment]
Any ideas? Anyone?

**Comment by rbgirshick** [comment]
A shot in the dark, but you may need to set dampening to 0 (see:
https://github.com/torch/optim/blob/master/sgd.lua#L35). Otherwise your LR
maybe be off by a factor of 10 (assuming momentum = 0.9) compared to Caffe.
Caffe's SGD update doesn't have dampening (or equivalently, dampening = 0).

**Comment by korkinof** [comment]
Wow! Many thanks! That wouldn't have crossed my mind!
I'll check it out tomorrow and let you know.

**Comment by korkinof** [comment]
So, the verdict is that the dampening was not the problem.
Mind you, I fixed some bugs that resulted in non-zero gradients ending up in the wrong positions, but the bias which I think was caused from the imbalance persisted.
Now, per-class normalisation of the gradient helped a looot and I got nice classification accuracy.
I am very curious if Caffe already does that, because I did not find anything inspecting the source code of the layers.

One of the things I do not understand is in the propotxt of the cls_loss:
layer {
  name: "rpn_loss_cls"
  type: "SoftmaxWithLoss"
  bottom: "rpn_cls_score_reshape"
  bottom: "rpn_labels"
  propagate_down: 1
  propagate_down: 0
  top: "rpn_cls_loss"
  loss_weight: 1
  loss_param {
    ignore_label: -1
    normalize: true
 }
What do the lines:
propagate_down: 1
propagate_down: 0
mean?

Just to reinforce my argument: I was trying out a random forest recently on the following Kaggle competition:
https://www.kaggle.com/c/santander-customer-satisfaction
The data are very imbalanced: 6% positives 94% negatives. When training on that, it gets everything to the negative class. If you apply weighting per category it gives some meaningful results.

**Comment by rbgirshick** [comment]
Thanks for the update. The caffe codebase doesn't do class balancing. It's
not surprising to hear that it helps and I'm glad you found that. However,
it should still work without rebalancing. It would be good to understanding
the underlying issue (i.e., there must still be an important difference
between your implementation and the reference one).

Re: propagate_down. Caffe usually does a good job of automatically figuring
out when to compute gradients wrt inputs, but in this case it didn't and
was trying to propagate gradients down to the RPN labels, which caused an
error. This flag manually forces it to not compute gradients wrt the second
input (rpn labels).

**Comment by korkinof** [comment]
Ok, that makes sense!
I'll keep looking into it for as much time as I can spare! There are indeed differences between my and the reference implementation and one of them must be causing this. I'll let you know how it goes!

**Comment by korkinof** [comment]
Just a small update, rpn regression error looks consistent with the one I'm getting from Torch.
I will run a full stage-1 in Caffe to see if there is a significant difference to Torch.
